### PR TITLE
Improve worksheet item manipulation commands

### LINKED ISSUE #201: Improve worksheet manipulation commands
cl add ^ -w worksheet

doesn't work and prints out

UsageError: bundle spec None doesn't match (want index 1 out of 0 bundles)

### PR DESCRIPTION
**Pull request by kashizui**
- Remove `cl wadd`
- Add only text items or subworksheets through `cl add`
- Create requirements file for running unit tests
- Fix test-cli issues

Fixes #201

@percyliang @pbhatnagar3

### CODE REVIEW

**Comment by kashizui** [comment]
cl help add

```
usage: cl add [-h] [-m MESSAGE [MESSAGE ...]] [-w WORKSHEET_SPEC]
              [subworksheet_spec [subworksheet_spec ...]]

Append worksheet links and/or text items to a worksheet.

positional arguments:
  subworksheet_spec     worksheets to add links to
                        ([<alias>::|<address>::]|(<uuid>|<name>))

optional arguments:
  -h, --help            show this help message and exit
  -m MESSAGE [MESSAGE ...], --message MESSAGE [MESSAGE ...]
                        add a text element (must appear after
                        subworksheet_specs)
  -w WORKSHEET_SPEC, --worksheet_spec WORKSHEET_SPEC
                        operate on this worksheet
                        ([<alias>::|<address>::]|(<uuid>|<name>))
```

File: codalab/lib/bundle_cli.py [hunk 1]
```diff
@@ -745,11 +744,11 @@ def do_download_command(self, args):
 
     @Commands.command(
         'cp',
-        help='Copy bundles across instances.',
+        help='Copy bundles across worksheets and instances.',
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Add that if between same instance, they just copying links.

**Comment by percyliang** [comment]
Code LGTM.

Something still bothers me about the interface (both naming and usage) though.  This is what we (would) have:
- cl wcp worksheet1 worksheet2: appends items from worksheet1 to worksheet2 (copying bundles across instances if necessary)
- cl cp bundle1 worksheet1: copies bundle to worksheet (copying across instances if necessary) - common case
- cl add worksheet1: adds text to current worksheet (doesn't go across instances)
- cl add -m text1: adds text to current worksheet

I think somehow this could be cleaner, given that we're willing to change the interface.  I think the followin would be more logical, although slightly more verbose (avoiding 'copying' is good I think, because most of the time, nothing actually gets copied):
- cl add bundle bundle1
- cl add worksheet worksheet1
- cl add text text1
- cl wadd worksheet1 worksheet2

**Comment by kashizui** [comment]
If we would do
`cl add worksheet worksheet1`
then what would be the function of wadd in this case (maybe you meant wcp)?

I like the explicit specification of item type, but your original problem
with `cl add bundle ^ -w worksheet` would still exist. Perhaps this is an
issue of documentation

**Comment by kashizui** [comment]
Or maybe we can require specification of the destination worksheet for `cl add` instead of a `-w` flag at all: in what case would we need an "operate completely on this worksheet" flag for `cl add`? We will typically be adding bundles from other worksheets, and this logic doesn't even make sense for adding subworksheets.

`cl add item_type item [item ...] dest_worksheet`

**Comment by kashizui** [comment]
- New syntax for cl add: `cl add item_type item [item ...]
  dest_worksheet`
- Update test-cli
- Include aliases in `cl help` output
- Clarify `cl cp` help

```
usage: cl add [-h] item_type item_spec [item_spec ...] dest_worksheet

Append text items, bundles, or subworksheets to a worksheet.

positional arguments:
  item_type       type of item(s) to add {text, bundle, worksheet}
  item_spec       item specifications, with the format depending on the specified item_type.
                      text:      (<text>|%<directive>)
                      bundle:    (<uuid>|<name>|^<index>)
                      worksheet: [<alias>::|<address>::]|(<uuid>|<name>)
  dest_worksheet  worksheet to which to add items [<alias>::|<address>::]|(<uuid>|<name>)

optional arguments:
  -h, --help      show this help message and exit
```

**Comment by percyliang** [comment]
I like putting dest_worksheet explicitly.

wadd is different because it adds the items in a worksheet to the
dest_worksheet.

**Comment by kashizui** [comment]
@percyliang I believe you're referring to `wcp`?

File: codalab/lib/bundle_cli.py [hunk 2]
```diff
@@ -502,6 +544,10 @@ def do_command(self, argv):
         argv = self.collapse_bare_command(argv)
         args = parser.parse_args(argv)
 
+        # print args.help
```
Comment (kashizui) [inline codalab/lib/bundle_cli.py]: remove

Comment (percyliang) [inline codalab/lib/bundle_cli.py]: I propose renaming wcp to wadd.

Comment (kashizui) [inline codalab/lib/bundle_cli.py]: Ah ok, makes sense to me.

**Comment by kashizui** [comment]
OK Next steps: remove `wcp` and `cp`

**Comment by kashizui** [comment]
New Interface

```
usage: cl add [-h] [-d] item_type item_spec [item_spec ...] dest_worksheet

Append text items, bundles, or subworksheets to a worksheet. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  item_type             type of item(s) to add {text, bundle, worksheet}
  item_spec             item specifications, with the format depending on the specified item_type.
                            text:      (<text>|%<directive>)
                            bundle:    ((<uuid>|<name>|^<index>)|(<alias>|<address>)::(<uuid>|<name>))
                            worksheet: [<alias>::|<address>::](<uuid>|<name>)
  dest_worksheet        worksheet to which to add items [<alias>::|<address>::](<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -d, --copy-dependencies
                        if adding bundles, will also add dependencies of the bundles if set
```

```
usage: cl wadd [-h] [-c] source_worksheet_spec dest_worksheet_spec

Append all the items on one worksheet to another. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  source_worksheet_spec
                        [<alias>::|<address>::]|(<uuid>|<name>)
  dest_worksheet_spec   [<alias>::|<address>::]|(<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -c, --clobber         clobber everything on destination worksheet with source (instead of appending)
```

@percyliang ready for review

File: codalab/lib/bundle_cli.py [hunk 3]
```diff
@@ -1699,15 +1702,15 @@ def do_wrm_command(self, args):
...
 
     @Commands.command(
-        'wcp',
-        help='Copy the contents from one worksheet to another.',
+        'wadd',
+        help='Append all the items on one worksheet to another. Bundles that do not yet exist on the destination service will be copied over.',
         arguments=(
             Commands.Argument('source_worksheet_spec', help=WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
             Commands.Argument('dest_worksheet_spec', help=WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
             Commands.Argument('-c', '--clobber', help='clobber everything on destination worksheet with source (instead of appending)', action='store_true'),
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Explicitly say not deleting the bundles on dest worksheet.
Maybe a better name is '--replace'?

**Comment by percyliang** [comment]
Nice - LGTM!  Remember to update the docs on the Wiki.

The only lingering hesitation that I have is that 'cl add worksheet foo .' is a bit verbose, but I think it's clear and I'll just get used to it.

**Comment by kashizui** [comment]
new `wadd` help:

```
usage: cl wadd [-h] [-r] source_worksheet_spec dest_worksheet_spec

Append all the items of the source worksheet to the destination worksheet, without modifying the existing items on the destination worksheet. Bundles that do not yet exist on the destination service will be copied over.

positional arguments:
  source_worksheet_spec
                        [<alias>::|<address>::](<uuid>|<name>)
  dest_worksheet_spec   [<alias>::|<address>::](<uuid>|<name>)

optional arguments:
  -h, --help            show this help message and exit
  -r, --replace         replace everything on the destination worksheet with the items from the source worksheet, instead of appending
```

**Comment by kashizui** [comment]
If we allow aliases for the item_type in the future, then people with strong CodaLab-fu could type short commands like `cl a w ^3 0x85817b`

**Comment by kashizui** [comment]
Updated wiki:
https://github.com/codalab/codalab/wiki/User_CodaLab-Worksheets-Tutorial

@percyliang could you update the executable papers tutorial (I think it has a mention of `wcp` at the bottom):
https://www.codalab.org/worksheets/0x9b6bb6fdd4aa4bf1b120300fd95f944a/

**Comment by percyliang** [comment]
Done.

**Pull request merged**